sdk/eventhub/azure-eventhub/azure/eventhub/_eventprocessor/ownership_manager.py

```diff
@@ -157,33 +138,64 @@ def _balance_ownership(self, ownership_list, all_partition_ids):
             active_ownership_by_owner[ownership["owner_id"]].append(ownership)
         active_ownership_self = active_ownership_by_owner[self.owner_id]
 
-        # calculate expected count per owner
+        # calculate expected and max count per owner
         all_partition_count = len(all_partition_ids)
         # owners_count is the number of active owners. If self.owner_id is not yet among the active owners,
         # then plus 1 to include self. This will make owners_count >= 1.
         owners_count = len(active_ownership_by_owner) + (
             0 if self.owner_id in active_ownership_by_owner else 1
         )
         expected_count_per_owner = all_partition_count // owners_count
+        # Py2 math.ceil() returns float, a/b return int if not divisable.
+        # Py3 math.ceil() returns int, a/b return float if not divisable.
+        max_count_per_owner = int(math.ceil(all_partition_count*1.0 / owners_count))
         # end of calculating expected count per owner
 
         to_claim = active_ownership_self
-        if len(active_ownership_self) < expected_count_per_owner:
-            # Either claims an inactive partition, or steals from other owners
-            if claimable_partition_ids:  # claim an inactive partition if there is
-                random_partition_id = random.choice(claimable_partition_ids)
-                random_chosen_to_claim = ownership_dict.get(
-                    random_partition_id,
-                    {
-                        "fully_qualified_namespace": self.fully_qualified_namespace,
-                        "partition_id": random_partition_id,
-                        "eventhub_name": self.eventhub_name,
-                        "consumer_group": self.consumer_group,
-                    },
+        if len(active_ownership_self) < max_count_per_owner:
+            to_try_steal = True
+            if (
+                LoadBalancingStrategy.GREEDY is self.load_balancing_strategy
+            ):  # Greedily claim more partitions if there are claimable partitions
+                to_greedy_claim_ids = random.sample(
+                    claimable_partition_ids, k=min(
+                        max_count_per_owner - len(active_ownership_self), len(claimable_partition_ids)
+                    )
                 )
-                random_chosen_to_claim["owner_id"] = self.owner_id
-                to_claim.append(random_chosen_to_claim)
-            else:  # steal from another owner that has the most count
+                if to_greedy_claim_ids:
+                    for pid in to_greedy_claim_ids:
+                        if pid in ownership_dict:
+                            to_claim_partition = ownership_dict[pid]
+                            to_claim_partition["owner_id"] = self.owner_id
+                            to_claim.append(to_claim_partition)
+                        else:
```

Comment:
I don't see this logic in the strategy of "# LoadBalancingStrategy.BALANCED"

For my understanding, I have two question for the case when a partition is already in its ownership_dict

1) does BALANCED strategy need to care about this case?
2) in the GREEDY strategy, we append the partitions which are already in ownership_dict into the to_claim list, does it mean these partitions would be re-claimed every load-balancing interval?

Reply:
Good catch. I wrote in a different way in two logic branches. That confused you. Updated the greedy branch to use the same way so the code is more readable.